core/src/main/java/jenkins/install/InstallUtil.java

```diff
@@ -204,6 +208,21 @@ private static InstallState getDefaultInstallState() {
         }
     }
 
+    private static boolean isEdgeCaseUpgrade() {
+        Jenkins j = Jenkins.getInstance();
+        // Edge case: used Jenkins 1 but did not save the system config page,
+        // the version is not persisted and returns 1.0, so try to check if
+        // they actually did anything
+        Map<String, TopLevelItem> items = new HashMap<String, TopLevelItem>(j.getItemMap());
+        for (Iterator<Map.Entry<String, TopLevelItem>> it = items.entrySet().iterator(); it.hasNext();) {
+            Map.Entry<String, TopLevelItem> entry = it.next();
+            if (new File(entry.getValue().getRootDir(), BUILT_IN_MARKER_FILE).exists()) {
+                it.remove();
+            }
+        }
+        return !items.isEmpty() || !j.getNodes().isEmpty();
```

Comment:
Actually nodes have a root dir but this is currently an implementation detail hidden in jenkins.model.Nodes, not sure how to expose it.